Add 2021 Roadmap for Clippy

flip1995 opened the pull request:
I'd like to write a roadmap for the Clippy repo and release it approximately at the same time as the Rust 2021 roadmap. This is a list of things I'd like to see happen in 2021 for Clippy. 

Before opening this up for everyone, I'd like the input of the team members. After we have some consensus in the team what we want to do, I will write this in a format similar to the [Rust roadmap](https://github.com/Mark-Simulacrum/rfcs/pull/1) and request comments from the broader community.

Feel free to comment about stuff you have in mind or comment what you don't like. This currently is just a semi-organized dump of ideas I have and every point needs to be written out in a separate tracking issue once we have consensus in the team and the community.

[Rendered](https://github.com/flip1995/rust-clippy/blob/roadmap/clippy2021.md)

Manishearth commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+can be important for sync-ups. I for example can't keep track of all the new
+lints and features that are implemented in Clippy, because I stopped looking at
+every PR a while ago. And no one should have to do this to be up-to-date.
+
```
Meetings might help. I'm not sure: i like asynchronous communication, and I feel like clippy is an ideal project for that to work since 90% of the stuff that happens is work on individual lints, and I largely feel it's _fine_ to not have a finger on the pulse for those. But as is glaringly obvious from this document, we have a _ton_ of large, high level projects we can do, and more synchronous communication would probably help them happen!

Writing this comment here not as a disagreement, really, just to create a point for discussion.

flip1995 replied:
> Writing this comment here not as a disagreement, really, just to create a point for discussion.

That's exactly what I wanted to achieve with this. What I wrote down is my perspective on things, where I would like to hear the opinion from others. 

---

As for the meetings: I think short meetings will help if we want to tackle the bigger tasks in this document. 

Additionally, we could do those for example bi-weekly before a Rust sync. That way people can raise concerns about things that should not be synced as-is or other things that are on their mind. Even when the meeting just consists of "no concerns" & "no updates from bigger tasks" and we end it after 2 minutes, that's fine. This might be especially helpful if we decide to grow the team, to get everyone on the same page.

Also I don't really get to look at issues anymore. So meetings like this can be helpful if someone wants to point out something we should discuss/address as a team instead of individuals. Or people that currently "only" do triage can bring up stuff where they see a trend in the issues.

=> If no one is against having regular meetings, I would like to try them out. If we find that they're not helpful, we can stop them again.

Mark-Simulacrum replied:
One thing I know for release team has definitely been true is that *having* a standing meeting, even if it's usually just 5 minutes or so, that you don't cancel even with an empty agenda can feel quite odd (what are we here for) but I personally find pretty useful. That meeting is weekly and scheduled for 30 minutes on Zulip, but it does mean that if we have some problem or issue, we can discuss at a known time. I find it can be helpful to just have a standing synchronous venue for some quick discussion.

(FWIW this sort of "do I run meetings? how do I run meetings?" work is part of the goals for the 2021 Rust roadmap charter work -- seeing the patterns across the project I think is going to be pretty helpful :)

flip1995 replied:
> FWIW this sort of "do I run meetings? how do I run meetings?" work is part of the goals for the 2021 Rust roadmap charter work

I know, that's where I got the idea for writing (some of) this. :)

phansch replied:
> I feel like clippy is an ideal project for that to work since 90% of the stuff that happens is work on individual lints

Hm, now I wonder if that's a result of asynchronous communication causing those tasks to be worked on more frequently than tasks that would require meetings first?

(I'm not opposed to either form of communication at all, just a train of thought :blue_heart:)

> If no one is against having regular meetings, I would like to try them out. 

I'm in favor of trying out meetings. Personally, I think they might also help build more of a stable contributor community around Clippy.

Manishearth replied:
> Hm, now I wonder if that's a result of asynchronous communication causing those tasks to be worked on more frequently than tasks that would require meetings first?

Yeah, that's kinda what I'm saying, for most stuff we don't need meetings, but we're large enough that it's now becoming easy to ignore the big ticket items. And we clearly have a lot of them!

Biweekly before a sync is actually kinda brilliant. The devtools meetings are monthly, text-based (i.e., on Discord, could use Zulip), and are relatively low effort.

Manishearth commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+### Management
+
+In 2020 we got over 1000 open issues with now between 25-35 open PRs. This is
```
From my perspective this is simultaneously a win and a loss 😄 . Clippy's maintainer set is particularly robust and can handle people dropping out pretty easily, so much that 3-4 people temporarily scaling back involvement did not actually cause the project to grind to a halt! But still, ideally we're as robust as possible!

Since most of our issues are lint suggestions I do not mind there being 1k lints 😄 , that's just 1k ideas for folks to implement. The PR backlog is worrying however.

Either way, I think it's _always_ a good idea to grow maintainership and processes. I do think there's a lot better we could do here, and excited to see what we can come up with!

flip1995 replied:
> Clippy's maintainer set is particularly robust and can handle people dropping out pretty easily

That is correct. This year was just a bit unlucky because besides the people who dropped out, the remaining maintainers (@Manishearth, @ebroto and I) had to deal with IRL stuff, which collided with hacktoberfest, which was unlucky. We're still dealing with the backlog, I think. More people would make things more stable, so I like to establish a process for growing the team.

> Since most of our issues are lint suggestions I do not mind there being 1k lints

Are they? I don't know. I have the feeling that bug reports are on the rise. This may also be something we want to look at. Which lint has how many issues? How many issues are new lint suggestions? Generally getting some metrics to track the state Clippy is in.

ebroto replied:
>This may also be something we want to look at. Which lint has how many issues? How many issues are new lint suggestions?

I agree. I also think we lack a process for deciding when a lint should be downgraded. How many false positives we accept for a warn/deny-by-default lint? And for a pedantic lint?

A good amount of our issues are quite old. IMO we should go through the whole issue list and re-triage, closing issues that don't apply anymore, and getting a better idea of lints that maybe should be downgraded.

Manishearth reviewed (approved):
This is _amazing_, thank you!

I like this clear outlook for the future, with focus not just on technical things, but on the maintenance/social aspect too.

I feel like the list of big ticket items is going to be a great way to grow more contributors: working on something like this is how @yaahc went from helping with random lints to working on a large, highly needed architectural change. It's going to be great to point growing contributors at this document as a way of saying "pick something if you want!".

Thanks so much for working on this ❤️

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+#### Lint groups
+
+The are more and more issues about managing lints in Clippy. Lints are hard to
+implement with a guarantee of no/few false positives (FPs). This won't change.
```
I think it would make sense to explain here that Clippy is less conservative than rustc in the kind of lints it accepts, as an explanation about why this is not going to change.

flip1995 replied:
Oh I rather meant, that implementing correct lints is hard and this fact won't change. But you're right, we should add a note, that Clippy is and will stay less conservative than rustc.

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+[#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+
+#### Clippy book
```
As stated in the linked issue, I think organizing (and improving) our current documentation about hacking on Clippy in a "Contributor Guide" like [cargo](https://doc.crates.io/contrib/) does would be helpful.

flip1995 replied:
I think writing a contributor guide in a book style would be even easier for us, since we already have the `doc/` documentation.

phansch replied:
I think putting both user-facing docs and contributing guides into the same book (but clearly separated) might help get some new maintainers on board as well.

phansch replied:
I think putting both user-facing docs and contributing guides into the same book (but clearly separated) might help get some new maintainers on board as well.

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+#### FP rate
+
+In the worst case, new lints are only available in nightly for 2 weeks, before
+hitting beta and ultimately stable. This and the fact that fewer people use
+nightly Rust nowadays makes it more more probably that a lint with many FPs hits
+stable. This leads to annoyed users, that will disable these new lints in the
+best case and to more annoyed users, that will stop using Clippy in the worst.
+We should think of a process to prevent this from happening. This will also be
+addressed later in this issue.
```
Seems to be repeated in both "Usability" and "Reliability" sections

flip1995 replied:
Yeah, this repeats 3 times pretty much. 2x external regarding usability (categories) & reliability (FP rate) and 1x internal regarding implementing a process to improve the situation. For me this is one of the things with the highest prio I want to address.

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+prioritize things. It is split up into external and internal issues, which are
+further split up.
+
+## External
```
Some other user facing stuff I would like to improve:
* Stabilizing `--fix`, and going through the lint list to see if we can turn more lints into `MachineApplicable`
* Support for multi-span suggestions in rustfix.

flip1995 replied:
Stabilizing fix goes with the `cargo check -> cargo clippy` stabilization, I think? Multi-span suggestions is out of scope for this document, because it is a rustfix issue.

ebroto replied:
> Stabilizing fix goes with the `cargo check -> cargo clippy` stabilization, I think? 

Yes, you're right.

>Multi-span suggestions is out of scope for this document, because it is a rustfix issue.

Although it's a rustfix issue, I think we are the more interested party in getting that to work.

matthiaskrgr commented:
I really like this. :+1: 
I got some more ideas about avoiding lint FPs from new lints riding up the release channels too quickly while reading this, will make a comment on that issue :D

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+#### More maintainer
+
+More people means less work for each individual. We should work out a process
+how onboarding of new folks can work and look out for people that are interested
+joining and taking up responsibility.
```
Maybe we could try to create "working groups" to address the bigger issues stated in this document. One or more team members could provide mentorship on the issue to people that wants to get involved in the project. That way 1) we help folks getting acquainted with the project 2) some of them can turn into new maintainers 3) we fix a long-standing issue.

Related to this, when this document is finished I think we should try to get it announced in the Rust blog so that it gets more attention.

flip1995 replied:
> Maybe we could try to create "working groups" to address the bigger issues stated in this document

I think for every bigger project in this document a team member should be assigned and see to get it over the finish line, either with mentorship or implementing it themself. If we do this, stand-up/sync-up meetings can help to track the state of the projects. But let's focus on "what" we want to do here and address the "How" later.

> Related to this, when this document is finished I think we should try to get it announced in the Rust blog so that it gets more attention.

Yes, as @Mark-Simulacrum wrote, we can also include it in the Rust roadmap document.

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+lints and features that are implemented in Clippy, because I stopped looking at
+every PR a while ago. And no one should have to do this to be up-to-date.
+
+### Development
```
Some other process I think we could improve is triaging.

* We should make sure all issues are triaged in a timely manner. As @flip1995 states in another comment, short sync meetings would help with that. We should make sure critical issues are going to be worked on.
* Ideally we will get help from contributors. It's another good way to get people involved, but it's not really an easy task for a newcomer to the project to do, so they should have direct support from the team.
* I think we should add a `triaged` tag. It does not happen often, but I would hate having to close a PR because the issue was not triaged and is not really an issue, or should not be fixed in Clippy, or...

Mark-Simulacrum commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+In 2020 we got over 1000 open issues with now between 25-35 open PRs. This is
+mostly because of the lack of resources in the team. According to the planned
+roadmap of the Rust project, we should improve our processes and look for more
+eyes and hands to help reviewing and triaging PRs and issues.
```
I would love to chat if you want about some of the attempts I've made historically on rust-lang/rust to deal with issues (though largely all of them have not been super successful, as our 6,000+ issues may show...). I know @jonas-schievink has been doing issue triage for the past year or two as well, too, and maybe something like the prioritization group process may be helpful too.

In general, this is an area where as Rust grows, we're going to want to work out better processes for dealing with issue influx. I don't know yet what those are, but I imagine that we're going to end up with the same problems (though maybe not the same solutions!) on many of our repositories :)

flip1995 replied:
I see, keeping this issue team-internal worked out well. You're just everywhere 😄 

But I would love to figure out things we can do across projects to improve triaging. For Clippy specifically, I wrote a small program before, which creates a list of issues sorted by lint (which never got anywhere). Another idea I had for PRs is to write a dashboard similar to the bors queue, which tracks the last activity on PRs, with information about the reviewer/author/waiting-on-{review,author} state.

Manishearth replied:
On triage: it's worth wondering if we can loop in the existing triage team to help. If we wrote a TRIAGING doc that lets others triage things here that would also be cool.

re: team-internal: sorry, i posted it in #core because i thought it was a really well designed roadmap, as a "this is cool", and particularly wanted Mark to have a look because he's been doing the main roadmap 😄

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+[#5394](https://github.com/rust-lang/rust-clippy/issues/5394)
+
+#### `rustc` integration
```
Some general issues I think we should address, that would involve some communication with the compiler team:

* Implement a solution for the problem of conflicting lints, probably directly in rustc?
* Clarify the situation with pre-expansion lints. This comes up quite often.

flip1995 replied:
Another thing that came to mind: Implementing the `#[expect(<lint>)]` attribute, which warns if the lint you expect to trigger doesn't trigger (anymore). That can help, if people allow lints because of FPs, but then never remove the ``allow`` even when the lint gets fixed.

flip1995 replied:
But I think this is also rather answering the "How" question, than the "What".

Mark-Simulacrum commented:
Just wanted to drop by and add that I'm super excited for folks to be doing this work, happy to see you've kicked it off here :heart: 

Do you mind if I include a link to this RFC (or if you plan to publish it on the main clippy repository?) in the Rust 2021 Roadmap RFC? I think it's great as a reference to the sort of individual team roadmap I was hoping for, and would be a great example for all of us :)

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+lints and features that are implemented in Clippy, because I stopped looking at
+every PR a while ago. And no one should have to do this to be up-to-date.
+
+### Development
```
Other big technical issues I think we should address:

* We depend on `compiletest-rs`, which I would say that nowadays is unmaintained. We also have the recurrent "multiple crates" issue because we don't have a good way to correctly specify dependencies. I've been looking into this and would like to start working towards a solution soon.
* Getting rid of hardcoded paths in favor of diagnostic items.

flip1995 replied:
I think there are many things we can improve here, I would like to have some but not too many examples in this document. (I would put this under the point "rustc integration")

ebroto commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+
+The are more and more issues about managing lints in Clippy. Lints are hard to
+implement with a guarantee of no/few false positives (FPs). This won't change.
+But we should consider implementing more lint groups or improve the process of
+classifying lints. This will also involve internal changes, and will be
```
IMO adding more granularity by splitting the lints in specific categories that are not directly related to the default warning level would help. I'm not sure if we can do that without breaking stability guarantees though.

flip1995 replied:
Let's discuss the "How" later.

ebroto reviewed (commented):
Thanks for doing this!

We need to organize to address the growing pains of the project, and I think this comes at the right time.

phansch commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+lint levels can be defined and reused across projects. Discussions about this
+often lead to nothing specific/actionable or to "we need an RFC for this". And
+this is exactly what we need to do. Get together with the cargo team and write
+an RFC and implement such a configuration file somehow and somewhere.
```
I'd LOVE it if this feature hits nightly in 2021!

AIUI, the 'only' thing left to figure out is how to ensure the reproducibility of compiling .crate files when lint configuration exists outside the project. There have been two proposed solutions to that problem in the tracking issue. On first glance, one of those (by rkuhn) seems plausible to implement. The other one (by jmaargh) would probably require a separate RFC and I'm not sure if it would be accepted just for the sake of the lint configuration feature. 
But yeah, this part definitely needs cargo team involvement.

phansch replied:
I can try to write up a summary in the tracking issue if that's alright

flip1995 replied:
I don't think we get far with only discussing this in issues. There were other concerns, like where to implement it: Should it be an extra file? Should it live in `Cargo.toml`? Or a hybrid version, so that you can link a file in the `Cargo.toml` but can also put it directly in there. Also the reproducibility of compiling probably needs more discussion.

I don't think this can be done without an RFC (in a joint effort of cargo+clippy).

phansch replied:
At least the matter of files seems to have been decided by the cargo team ([see this IRLO comment](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135/6)) to be `Cargo.toml`

But yes, an RFC is absolutely needed =)

flip1995 replied:
> At least the matter of files seems to have been decided by the cargo team

I don't know, if that's really decided. I think the cargo team would really like for it to go into `Cargo.toml` but most people find a extra file for this more intuitive. That's at least the impression I got from closing the issues about this feature in the `clippy.toml` file and other people I talked to about this.

flip1995 commented:
> I got some more ideas about avoiding lint FPs from new lints riding up the release channels too quickly while reading this, will make a comment on that issue :D

@matthiaskrgr That's great! Keep in mind that I want to keep the scope of this document on what we want to do and figure out the "How" in dedicated issues.

> Do you mind if I include a link to this RFC (or if you plan to publish it on the main clippy repository?) in the Rust 2021 Roadmap RFC?

@Mark-Simulacrum Sure! But please wait until we publish it. Maybe we can coordinate and publish it at the same time as the Rust roadmap. Do you already know a date when you want to publish it?

phansch commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+2. Make `cargo dev` work in the Rust repo, just as it works in the Clippy repo.
+   E.g. `cargo dev bless` or `cargo dev update_lints`. And even add more things
+   to it that might be useful for the Rust repo, e.g. `cargo dev deprecate`.
+3. Easier sync process. The `subtree` situation is not ideal.
```
> The `subtree` situation is not ideal.

We still require the manual patch for `subtree` right?

flip1995 replied:
Yes and the first time running `git subtree` still takes about 1h. And with every additional commit it takes longer.

The later, we can solve by a "re-sync", but that results in every synced commit appearing twice in `git log`.

phansch commented on clippy2021.md:
```diff
@@ -0,0 +1,178 @@
+As already mentioned above, classifying new lints gets quite hard, because the
+probability of a buggy lint getting into stable is quite high. We should work
+out a process (and maybe tooling) to make sure we test lints, before releasing
+them into the wild.
```
I agree, I think we should be much more careful before enabling new lints by default. We definitely need a tool or good instructions to make testing new lints on _real-world_ code easier. From my experience, the harder it is to test something, the less likely it is to be tested properly at all. A quick feedback loop is very important, imo.

(Also see [this comment here](https://github.com/rust-lang/rust-clippy/issues/6429#issuecomment-741153345) about cppcheck testing)

Mark-Simulacrum commented:
I am planning on posting the main roadmap on the 17th.

flip1995 committed:
Add Roadmap for 2021

flip1995 commented:
@Manishearth @ebroto @phansch @matthiaskrgr @Mark-Simulacrum I pushed the re-written Roadmap document: [Rendered](https://github.com/flip1995/rust-clippy/blob/roadmap/doc/roadmap-2021.md). This is also the version I want to publish.

Some open questions before publishing:

- Should we write a blog post after publishing it?
- I think I addressed all comments. Please re-review :)
- I think I had something else, but it is late and I'm tired and forgot. Probably wasn't important :sweat_smile:

phansch reviewed (approved):
> Should we write a blog post after publishing it?

Yes, I think more feedback will help us with prioritization a lot, especially since we don't have the call for blog posts like the main roadmap.

> I think I addressed all comments. Please re-review :)

:+1: from my side

Manishearth commented:
+1 from me, I'm okay with a blog post if folks wanna do it. I'm excited!

Mark-Simulacrum commented:
FYI: My plan is to edit in a link to this roadmap after publishing the initial PR of mine, just to avoid any need for coordinating that to happen at similar times or anything.

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+
+Members of the Clippy team will be assigned tasks from one or more of these
+topics. The team member is then responsible to complete the assigned tasks. This
+can either be done by implementing it or by providing mentorship to interested
```
```suggestion
can either be done by implementing them or by providing mentorship to interested
```

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+
+# Motivation
+
+With the ongoing growth of the Rust language and with that the whole ecosystem,
+also Clippy gets more and more users and contributors. This is good for the
```
```suggestion
With the ongoing growth of the Rust language and with that of the whole ecosystem, Clippy gets more and more users and contributors. This is good for the
```

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+
+Additionally, according to the [Rust Roadmap 2021], clear processes should be
+defined by every team and unified across teams. This Roadmap is the first step
+for doing this.
```
```suggestion
towards this.
```

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+
+#### No Output After `cargo check`
+
+Currently when `clippy` is run after `cargo check`, it does not produce any
```
```suggestion
Currently when `cargo clippy` is run after `cargo check`, it does not produce any
```
?

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+#### Scaling up the Team
+
+More people means less work for each individual. Together with the document
+about expectations for team member, a document defining the process of how to
```
```suggestion
about expectations for team members, a document defining the process of how to
```

matthiaskrgr commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+repository. This made syncing between the two repositories easier. A
+`#[non_exhaustive]` list of things that still can be improved is:
+
+1. Use the same `rustfmt` and configuration as `rustc`.
```
```suggestion
1. Use the same `rustfmt` version and configuration as `rustc`.
```

flip1995 commented on doc/roadmap-2021.md:
```diff
@@ -0,0 +1,228 @@
+
+# Motivation
+
+With the ongoing growth of the Rust language and with that the whole ecosystem,
+also Clippy gets more and more users and contributors. This is good for the
+project, but also brings challenges along. Some of these challenges are:
```
```suggestion
With the ongoing growth of the Rust language and with that of the whole
ecosystem, also Clippy gets more and more users and contributors. This is good
for the project, but also brings challenges along. Some of these challenges are:
```

flip1995 commented:
This should be ready to get published now. I want to wait for the Rust roadmap before publishing this though, since I reference it quite a bit.

Manishearth commented:
@flip1995 the rust roadmap is already a PR: https://github.com/rust-lang/rfcs/pull/3037, i think it's fine to merge this first since it will want to refer to us too.

flip1995 commented:
Thanks for notifying me! I'll open a PR to the Clippy repo in a bit (~1h). I want to get community feedback for this before actually merging it in the main repo.

(you can also open a PR for me, if you want)

flip1995 commented:
cc https://github.com/rust-lang/rust-clippy/pull/6462
